doc/developer_guild.md

```diff
@@ -0,0 +1,471 @@
+
+```cpp
+const char *MM(Tensor a, Tensor b, Tensor *result) {
+  try {
+    at::Tensor c = at::mm(*a, *b);
+    *result = new at::Tensor(c);
+    return nullptr;
+  } catch (const std::exception &e) {
+    return exception_str(e.what());
+  }
+}
+```
+
+It runs the following steps, which are in most other wrappers too.
+
+1. It calls the native function `mm` and creates the result tensor `c` on the
+   stack.
+1. It allocates a heap object `*result` and **"copies"** `c` to `*result`.  This
+   step is necessary because the return from `MM` will destruct `c`.  This step
+   is highly efficient as it doesn't actually copy the content of `c`, because
+   `at::Tensor` contains only a smart pointer to the underlying tensor content.
+1. It returns the string-serialization of the exception if there is any, or
+   `nullptr`.
+
+### Go Wrappers and Go Tensor
+
+In package `gotorch`, we define the Go wrappers of native functions, which
+operates the Go type `Tensor` defined in `tensor.go`.
+
+```go
+// Tensor wrappers a pointer to C.Tensor
+type Tensor struct {
+    T *unsafe.Pointer
+}
+```
+
+At first glance, this definition looks too much more complicated than the C type
+Tensor, which is a pointer.
+
+The literal translation of the C type Tensor into Go is
+
+```go
+type Tensor unsafe.Pointer
+```
+
+The Go type `unsafe.Pointer` represents any C pointer type.
+
+However, this is not enough because we need to attach to each `Tensor` a
+[*finalizer*](https://golang.org/pkg/runtime/#SetFinalizer) to free the
+underlying `at::Tensor` object when a Go tensor is out-of-use.  Only Go pointers
+but not C pointers could have finalizers attached, so we add a `*`, indicating a
+Go pointer, to the above definition, this makes it a Go pointer to a C pointer.
+
+```go
+type Tensor *unsafe.Pointer
+```
+
+This pointer-to-pointer is still not enough as we need to attach methods of
+`at::Tensor`, like `MM`, `Add`, and `To`, to the Go type.  Again, Go pointers
+cannot have methods, so we wrap the above pointer-to-pointer in a struct.
+
+Given the Go type `Tensor`, the Go wrapper `MM` is as follows.
+
+```go
+package gotorch
+
+func MM(a, b Tensor) Tensor {
+    var t C.Tensor
+    MustNil(unsafe.Pointer(C.MM(C.Tensor(*a.T), C.Tensor(*b.T), &t)))
+    SetTensorFinalizer((*unsafe.Pointer)(&t))
+    return Tensor{(*unsafe.Pointer)(&t)}
+}
+```
+
+It runs the following steps:
+
+1. It passes the address of `t` of type `C.Tensor` to the C wrapper, `C.MM`.
+   Because `C.Tensor` is a pointer to `at::Tensor`, the line `*result = new
+   at::Tensor(c)` in the C wrapper makes `t` pointing to the newly allocated
+   tensor on the heap.
+1. It calls `MustNil` to check the C string returned from the C wrapper.
+   `MustNil` panics if the string is not `nullptr`.
+1. It calls `gotorch.SetTensorFinalizer` to attach the finalizer to the tensor
+   returned by `C.MM`.
+1. It returns a value of the Go type `Tensor` that encapsulates `t`.
+
+## Define GoTorch Functionals
```

Comment:
Let us do not remove the existing file. But add a new one --  Porting PyTorch Functions and Modules

Reply:
Done.